packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -232,10 +236,16 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
       if (allowsCustomValue && inputValue !== itemText) {
         commitCustomValue();
       } else {
-        resetInputValue();
-        // Close menu if blurring away from the combobox
-        // Specifically handles case where user clicks away from the field
-        triggerState.close();
+        // If selected key, open state, and inputvalue are controlled, fire onSelectionChange so that application can control
+        // closing the menu when blurring away from the combobox. Otherwise reset the input value and close the menu ourselves
+        if (props.selectedKey !== undefined && props.isOpen !== undefined && props.inputValue !== undefined) {
+          props.onSelectionChange(selectedKey);
+        } else {
+          resetInputValue();
+          // Close menu if blurring away from the combobox
+          // Specifically handles case where user clicks away from the field
+          triggerState.close();
```

Comment:
a stopgap solution to fix the following flow:
1. type an invalid value in the "inputValue, selectedKey, and isOpen (controlled)" story (e.g. type "z")
2. blur away from the combobox
3. The menu opens (it shouldn't)

The above happens because the `onInputChange` that occurs from calling `resetInputValue` happens after `triggerState.close()` for some reason, still digging into why that is

Reply:
After digging, I think reason this bug happens in the first place is that the state updates from `resetInputValue` and `triggerState.close()` are batched. Here is a run down:
1. User types in a invalid value and blurs the field. The state at this point is `{isOpen: false, inputValue: "z"}`
2. This causes `resetInputValue` to trigger which will trigger `onInputChange` with `{isOpen: true, inputValue: ""}` as the new state to be set. This state change isn't made just yet.
3. `triggerState.close()` is triggered which attempts a state update of `{isOpen: false}`. Since the `onInputChange` from `resetInputValue` hasn't happened since they will be batched, the current `isOpen` state is still `false`. This means `onOpenChange` doesn't get called by `useControlledState` since there isn't a change in the open state.
4. The end result is `onInputChange` fires with `{isOpen: false, inputValue: ""}`  but `onOpenChange` doesn't fire at all, resulting in the menu being opened after the blur

One way to fix this is by forcing the state updates from `resetInputValue` and `triggerState.close()` to not batch by wrapping it in a `Promise.resolve().then`. Bleh.

An alternative would be to manually call `props.onOpenChange()` for this specific case.
```
          resetInputValue();
          if (!triggerState.isOpen) {
            props.onOpenChange(false);
          } else {
            triggerState.close();
          }
```

Reply:
Another alternative would be to have the following in the render `useEffect`:
```
    if (!isFocused && triggerState.isOpen) {
      triggerState.close();
    }
```
This would force the menu close if the combobox isn't focused. Its a bit restrictive since it would mean end users wouldn't be able to fully control the open state of the combobox when it isn't focused but that might be ok.

Reply:
@snowystinger suggested that perhaps the user should track whether or not the combobox is currently focused or not and use that information to better inform `onInputChange` as to if it should set isOpen to true or false. Tentatively, maybe pass `isFocused` to `onInputChange` like so:
```
  let informedOnInputChange = (value) => {
    props.onInputChange(value, isFocused)
  }
  let [inputValue, setInputValue] = useControlledState(
    props.inputValue,
    props.defaultInputValue ?? '',
    informedOnInputChange
  );
```
This might require that we track focus state via a ref as well so that we are up to date.